### PR TITLE
Fix optional outputs support

### PR DESCRIPTION
**Pull request by rogday**
https://github.com/onnx/onnx/blob/main/docs/IR.md#optional-inputs-and-outputs

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -678,10 +678,26 @@ void ONNXImporter::populateNet()
     CV_LOG_DEBUG(NULL, "DNN/ONNX: import completed!");
 }
 
+const std::string& extractNodeName(const opencv_onnx::NodeProto& node_proto)
+{
+    for (int i = 0; i < node_proto.output_size(); ++i)
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:683]: IMHO, we should try to use this value first (before outputs):

https://github.com/opencv/opencv/blob/22f0bcaf8f4fb406ae251830cf99b7ea05de40e4/modules/dnn/src/onnx/opencv-onnx.proto#L165

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -802,6 +830,7 @@ void ONNXImporter::parseReduce(LayerParams& layerParams, const opencv_onnx::Node
 {
     opencv_onnx::NodeProto node_proto = node_proto_;
     const std::string& layer_type = node_proto.op_type();
+    const std::string output_name = node_proto.output(0);
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:833]: Reference?

`const std::string&`

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:833]: `node_proto`'s output name is modified below.

**Review by alalek — APPROVED** [review]
LGTM :+1:

**Comment by alalek** [comment]
It makes sense to test such non-trivial changes with 4.x branch too before merging them.

Currently we have broken output names with failed tests on 4.x branch:

```
[  FAILED  ] Objdetect_face_detection.regression
[  FAILED  ] Objdetect_face_recognition.regression
```

Outputs are also broken on 3.4 branch, but we don't have tests for them here.

**Pull request merged**